optimization.py

```diff
@@ -130,10 +130,11 @@ def mutate(vec):
     # 1. step not in range check
```

Comment:
Thanks! I think the `XXX: broken.` in line 129 means that I think that this function is generally wrong. Note that the code is from http://www.amazon.com/Programming-Collective-Intelligence-Building-Applications/dp/0596529325 , I didn't write it myself.

Can you explain what problem you are fixing? Maybe you could add a test to optimization_test.py that fails without this change but passes with it?